src/iceberg/test/avro_test.cc

```diff
@@ -244,4 +246,234 @@ TEST_F(AvroReaderTest, AvroWriterNestedType) {
+  auto schema = std::make_shared<iceberg::Schema>(std::vector<SchemaField>{
+      SchemaField::MakeRequired(1, "id", std::make_shared<IntType>()),
+      SchemaField::MakeOptional(2, "name", std::make_shared<StringType>()),
+      SchemaField::MakeRequired(
+          3, "nested",
+          std::make_shared<iceberg::StructType>(std::vector<SchemaField>{
+              SchemaField::MakeRequired(4, "value", std::make_shared<DoubleType>())}))});
+
+  std::string expected_string = R"([
+    [1, "Alice", [3.14]],
+    [2, null, [2.71]],
+    [3, "Bob", [1.41]]
+  ])";
+
+  // Test with direct decoder (default)
+  {
+    temp_avro_file_ = CreateNewTempFilePathWithSuffix(".avro");
+    WriteAndVerify(schema, expected_string);
+  }
+
+  // Test with GenericDatum decoder
+  {
+    temp_avro_file_ = CreateNewTempFilePathWithSuffix("_generic.avro");
+    auto reader_properties = ReaderProperties::default_properties();
+    reader_properties->Set(ReaderProperties::kAvroUseDirectDecoder, false);
+
+    ArrowSchema arrow_c_schema;
+    ASSERT_THAT(ToArrowSchema(*schema, &arrow_c_schema), IsOk());
+    auto arrow_schema_result = ::arrow::ImportType(&arrow_c_schema);
+    ASSERT_TRUE(arrow_schema_result.ok());
+    auto arrow_schema = arrow_schema_result.ValueOrDie();
+
+    auto array_result = ::arrow::json::ArrayFromJSONString(arrow_schema, expected_string);
+    ASSERT_TRUE(array_result.ok());
+    auto array = array_result.ValueOrDie();
+
+    struct ArrowArray arrow_array;
+    auto export_result = ::arrow::ExportArray(*array, &arrow_array);
+    ASSERT_TRUE(export_result.ok());
+
+    std::unordered_map<std::string, std::string> metadata = {{"k1", "v1"}};
+
+    auto writer_result =
+        WriterFactoryRegistry::Open(FileFormatType::kAvro, {.path = temp_avro_file_,
+                                                            .schema = schema,
+                                                            .io = file_io_,
+                                                            .metadata = metadata});
+    ASSERT_TRUE(writer_result.has_value());
+    auto writer = std::move(writer_result.value());
+    ASSERT_THAT(writer->Write(&arrow_array), IsOk());
+    ASSERT_THAT(writer->Close(), IsOk());
+
+    auto file_info_result = local_fs_->GetFileInfo(temp_avro_file_);
+    ASSERT_TRUE(file_info_result.ok());
+
+    auto reader_result = ReaderFactoryRegistry::Open(
+        FileFormatType::kAvro, {.path = temp_avro_file_,
+                                .length = file_info_result->size(),
+                                .io = file_io_,
+                                .projection = schema,
+                                .properties = std::move(reader_properties)});
+    ASSERT_THAT(reader_result, IsOk());
+    auto reader = std::move(reader_result.value());
+    ASSERT_NO_FATAL_FAILURE(VerifyNextBatch(*reader, expected_string));
+    ASSERT_NO_FATAL_FAILURE(VerifyExhausted(*reader));
+  }
+}
+
+TEST_F(AvroReaderTest, LargeDataset) {
```

Comment:
It seems that we don't have test for projection. For example, only some columns are selected and they are reordered.